src/utils/signalsmooth.py

```diff
@@ -49,137 +49,212 @@ def __call__(self, data, pad_value=0, **kwargs):
-    y = np.convolve(w/w.sum(), s, mode='same')
-    return y[window_len-1:-window_len+1]
-
+    
+    if window == 'flat': #moving average
+        w = np.ones(window_len,'d')
+    else:
+        w = getattr(np, window)(window_len)
+    y = np.convolve(w/w.sum(), s, mode='same')
+    return y[window_len//2:-(window_len//2)]
+
 
 #*********** part2: 2d
 
 from scipy import signal
 
-def gauss_kern(size, sizey=None):
-    """ Returns a normalized 2D gauss kernel array for convolutions """
-    size = int(size)
-    if not sizey:
-        sizey = size
-    else:
-        sizey = int(sizey)
-    x, y = np.mgrid[-size:size+1, -sizey:sizey+1]
-    g = np.exp(-(x**2/float(size) + y**2/float(sizey)))
-    return g / g.sum()
-
-@UnmaskAndPad
-def smooth_image(im, n=10, ny=None, mode='valid') :
-    """ blurs the image by convolving with a gaussian kernel of typical
-        size n. The optional keyword argument ny allows for a different
-        size in the y direction.
-    """
-    g = gauss_kern(n, sizey=ny)
-    improc = signal.convolve(im, g, mode=mode)
-    return(improc)
-
-
+def twoDim_kern(size, window, sizey=None):
+    """ Returns a normalized 2D kernel array for convolutions """    
+    size = int(size)
+    if not sizey:
+        sizey = size
+    else:
+        sizey = int(sizey)    
+    x, y = np.mgrid[-size:size+1, -sizey:sizey+1]    
+    if window=='gauss':
+        g = np.exp(-(x**2/float(size) + y**2/float(sizey)))
+    elif window=='flat':
+        g = np.ones((size,sizey))
+    elif window=='hanning':
+        g1d_x = np.hanning(size)
+        g1d_y = np.hanning(sizey)        
+        g = np.sqrt(np.outer(g1d_x,g1d_y))
+    elif window=='hamming':
+        g1d_x = np.hamming(size)
+        g1d_y = np.hamming(sizey)        
+        g = np.sqrt(np.outer(g1d_x,g1d_y))
+    elif window=='bartlett':
+        g1d_x = np.bartlett(size)
+        g1d_y = np.bartlett(sizey)        
+        g = np.sqrt(np.outer(g1d_x,g1d_y))
+    elif window=='blackman':    
+        g1d_x = np.blackman(size)
+        g1d_y = np.blackman(sizey)        
+        Temp = np.outer(g1d_x,g1d_y)
+        Temp[np.abs(Temp)<1e-15] = 0        
+        g = np.sqrt(Temp) 
+        # NOTE: For the blackman window some elements have tiny negative values which
+        #   become problematic when taking the square root. So I've added the above
+        #   code to fix this.
+    return g/g.sum()
+
+@UnmaskAndPad
+def smooth_image(im, window='gauss', n=10, ny=None):
+    """ blurs the image by convolving with a kernel of typical
+        size n. The optional keyword argument ny allows for a different
+        size in the y direction.
+    """
+    n = int(n)
+    if not ny:
+        ny = n
+    else:
+        ny = int(ny)
+    g = twoDim_kern(size=n,window=window,sizey=ny) 
+    [mx,my] = im.shape
+    ox = 2*(n//2)+mx
+    oy = 2*(ny//2)+my
+    S = np.zeros((ox,oy))
+    S[n//2:-(n//2),ny//2:-(ny//2)] = im
+    for i in np.arange(n//2,ox-(n//2)):

```

Comment:
I am not convinced that the loop is actually necessary here - I feel like this could be done with array/vector statements only (you would have to choose the index range for S properly) - but this is not a big deal, since the inner dimension is already vectorized and this is not performance-critical